### ISSUE TITLE
[LightGBM] [Fatal] Length of labels differs from the length of #data when working with parquet data.

### ISSUE THREAD

**Issue by matteociccozzi-ah** [microsoft/LightGBM#6920]
## Description

When using DaskLGBMClassifier to train on a Dask DataFrame loaded from multiple Parquet files, the fit call fails unless the DataFrame is explicitly persisted with data = data.persist() (and wait(data)) before splitting out features and target. This requirement contradicts the Parallel Learning Guide, which states that LightGBM should automatically concatenate and align partitions on each Dask worker without needing a prior manual persist(). In the reproducible example, commenting out the persist/wait block triggers an exception or misalignment during clf.fit(), whereas including it works as expected.

**Error:**
```
[LightGBM] [Fatal] Length of labels differs from the length of #data
2025-05-20 14:29:45,067 - distributed.worker - ERROR - Compute Failed
Key:       _train_part-c95b95b2-32f1-40cc-b143-6147273a9246
State:     executing
Task:  <Task '_train_part-c95b95b2-32f1-40cc-b143-6147273a9246' _train_part(, ...)>
Exception: "LightGBMError('Length of labels differs from the length of #data')"
Traceback: '  File "/Users/mciccozzi/Desktop/mlops/models/model-lightgbm/.venv/lib/python3.11/site-packages/lightgbm/dask.py", line 340, in _train_part\n    model.fit(\n  File "/Users/mciccozzi/Desktop/mlops/models/model-lightgbm/.venv/lib/python3.11/site-packages/lightgbm/sklearn.py", line 1560, in fit\n    super().fit(\n  File "/Users/mciccozzi/Desktop/mlops/models/model-lightgbm/.venv/lib/python3.11/site-packages/lightgbm/sklearn.py", line 1049, in fit\n    self._Booster = train(\n                    ^^^^^^\n  File "/Users/mciccozzi/Desktop/mlops/models/model-lightgbm/.venv/lib/python3.11/site-packages/lightgbm/engine.py", line 297, in train\n    booster = Booster(params=params, train_set=train_set)\n              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File "/Users/mciccozzi/Desktop/mlops/models/model-lightgbm/.venv/lib/python3.11/site-packages/lightgbm/basic.py", line 3656, in __init__\n    train_set.construct()\n  File "/Users/mciccozzi/Desktop/mlops/models/model-lightgbm/.venv/lib/python3.11/site-packages/lightgbm/basic.py", line 2590, in construct\n    self._lazy_init(\n  File "/Users/mciccozzi/Desktop/mlops/models/model-lightgbm/.venv/lib/python3.11/site-packages/lightgbm/basic.py", line 2209, in _lazy_init\n    self.set_label(label)\n  File "/Users/mciccozzi/Desktop/mlops/models/model-lightgbm/.venv/lib/python3.11/site-packages/lightgbm/basic.py", line 3078, in set_label\n    self.set_field("label", label_array)\n  File "/Users/mciccozzi/Desktop/mlops/models/model-lightgbm/.venv/lib/python3.11/site-packages/lightgbm/basic.py", line 2847, in set_field\n    _safe_call(\n  File "/Users/mciccozzi/Desktop/mlops/models/model-lightgbm/.venv/lib/python3.11/site-packages/lightgbm/basic.py", line 313, in _safe_call\n    raise LightGBMError(_LIB.LGBM_GetLastError().decode("utf-8"))\n'

```
## Reproducible example

```
def example():
    import dask.dataframe as dd
    from dask.distributed import Client, LocalCluster
    import os
    import pandas as pd
    from distributed import wait
    from sklearn.datasets import make_classification
    from sklearn.metrics import accuracy_score
    from lightgbm import DaskLGBMClassifier
    
    X, y = make_classification(n_samples=100000, n_features=20, n_informative=2, n_redundant=10)
    df = pd.DataFrame(X, columns=[f"feature_{i}" for i in range(20)])
    df['target'] = y

    df_size = len(df)
    chunk_size = df_size // 2

    os.makedirs("tmp", exist_ok=True)

    df.iloc[:chunk_size].to_parquet(f"tmp/part-0.parquet")
    df.iloc[chunk_size:].to_parquet(f"tmp/part-1.parquet")


    # Setup a local Dask cluster
    cluster = LocalCluster(n_workers=2, memory_limit='2GB')
    client = Client(cluster)
    print(f"Dask client dashboard link: {client.dashboard_link}")

    data_path = "tmp"

    try:
        print(f"Attempting to read parquet files from: {data_path}")

        local_files = []
        for f in os.listdir(data_path):
            file_path = os.path.join(data_path, f)
            if file_path.endswith(".parquet"):
                local_files.append(file_path)

        data = dd.read_parquet(local_files)

        # REMOVING COMMENT BELOW FIXED THE ISSUE
        # data = data.persist()
        # _ = wait(data)
        
        x_train, y_train = data.drop(columns=["target"]), data["target"]

        clf = DaskLGBMClassifier()
        clf.fit(x_train, y_train)

        preds = clf.predict(x_train)
        preds_local = preds.compute()
        actuals_local = y_train.compute()
        acc = accuracy_score(actuals_local, preds_local)

        print(f"ACCURACY: {acc}")
        
    except Exception as e:
        print(f"An error occurred: {e}")
    finally:
        # Shutdown the Dask client and cluster
        print("\nShutting down Dask client and cluster...")
        client.close()
        cluster.close()
        print("Dask cluster shut down.")


if __name__ == "__main__":
    example()
```

## Environment info

LightGBM version or commit hash:
```
name = "lightgbm"
version = "4.6.0"
source = { registry = "https://artifactory.dev.earnin.net/artifactory/api/pypi/pypi-virtual/simple" }
dependencies = [
    { name = "numpy" },
    { name = "scipy" },
]
sdist = { url = "https://artifactory.dev.earnin.net/artifactory/api/pypi/pypi-virtual/packages/packages/68/0b/a2e9f5c5da7ef047cc60cef37f86185088845e8433e54d2e7ed439cce8a3/lightgbm-4.6.0.tar.gz", hash = "sha256:cb1c59720eb569389c0ba74d14f52351b573af489f230032a1c9f314f8bab7fe" }
wheels = [
    { url = "https://artifactory.dev.earnin.net/artifactory/api/pypi/pypi-virtual/packages/packages/f2/75/cffc9962cca296bc5536896b7e65b4a7cdeb8db208e71b9c0133c08f8f7e/lightgbm-4.6.0-py3-none-macosx_10_15_x86_64.whl", hash = "sha256:b7a393de8a334d5c8e490df91270f0763f83f959574d504c7ccb9eee4aef70ed" },
    { url = "https://artifactory.dev.earnin.net/artifactory/api/pypi/pypi-virtual/packages/packages/21/1b/550ee378512b78847930f5d74228ca1fdba2a7fbdeaac9aeccc085b0e257/lightgbm-4.6.0-py3-none-macosx_12_0_arm64.whl", hash = "sha256:2dafd98d4e02b844ceb0b61450a660681076b1ea6c7adb8c566dfd66832aafad" },
    { url = "https://artifactory.dev.earnin.net/artifactory/api/pypi/pypi-virtual/packages/packages/64/41/4fbde2c3d29e25ee7c41d87df2f2e5eda65b431ee154d4d462c31041846c/lightgbm-4.6.0-py3-none-manylinux2014_aarch64.whl", hash = "sha256:4d68712bbd2b57a0b14390cbf9376c1d5ed773fa2e71e099cac588703b590336" },
    { url = "https://artifactory.dev.earnin.net/artifactory/api/pypi/pypi-virtual/packages/packages/42/86/dabda8fbcb1b00bcfb0003c3776e8ade1aa7b413dff0a2c08f457dace22f/lightgbm-4.6.0-py3-none-manylinux_2_28_x86_64.whl", hash = "sha256:cb19b5afea55b5b61cbb2131095f50538bd608a00655f23ad5d25ae3e3bf1c8d" },
    { url = "https://artifactory.dev.earnin.net/artifactory/api/pypi/pypi-virtual/packages/packages/5e/23/f8b28ca248bb629b9e08f877dd2965d1994e1674a03d67cd10c5246da248/lightgbm-4.6.0-py3-none-win_amd64.whl", hash = "sha256:37089ee95664b6550a7189d887dbf098e3eadab03537e411f52c63c121e3ba4b" },
]
```

Command(s) you used to install LightGBM

```shell
uv add lightgbm
```

python version: 3.11.11

## Additional Comments

I was expecting lightgbm to be able to align the label and features when setting up the training data like described here: https://lightgbm.readthedocs.io/en/latest/Parallel-Learning-Guide.html#setting-up-training-data

However, seems that is not the case when I tried on my parquet datasets. I had to add an explicit: 

```
        data = data.persist()
        _ = wait(data)

```
before proceeding to select my features and target.

**Comment by matteociccozzi-ah** [comment]
@jameslamb Sorry to call you out directly, do you have any idea of why this could be happening? I'm open to debugging myself, just not sure where to start looking so any advice is greatly appreciated!

**Comment by kol-di** [comment]
I can confirm precisely the same issue for package versions below
```
dask==2024.8.0 
lightgbm==4.6.0
```
Doing data.persist() indeed helps, but I would be really grateful to get some other solution, since the latter doesn't scale well

**Comment by jameslamb** [comment]
> [@jameslamb](https://github.com/jameslamb) Sorry to call you out directly, do you have any idea of why this could be happening? I'm open to debugging myself, just not sure where to start looking so any advice is greatly appreciated!

Sorry, it will be a while until I'm able to look into this. @jmoralez is very familiar with Dask and `lightgbm.dask`, he might be able to help.

One place where you could help improve the project AND help us make progress on this report is by improving that error message. It comes from here:

https://github.com/microsoft/LightGBM/blob/e7c6c4371b5d725902a09a80b4d6c36e432a4381/src/io/metadata.cpp#L422

The error message would be more helpful if it reported the actual sizes, like:

```text
'Length of labels (1000) differs from the number of rows (1500)'
```

Comparing those sizes to the partition sizes might help us to narrow down where this problem comes from.

I'd welcome a PR to make that change.
